The `Repository` - New way of handling collections and Cloud synchronization with optimistic updates.

holoisme opened the pull request:
Hello,

As you might have noticed in the main app, a lot of the CRUD operations (Create, Read, Update and Delete) are slow and feel clunky. This is because every operation goes as follows : First we use `callFunction` to update the state of the collections in the cloud, then the UI waits changes using variants of `useDynamicDocs`.
This is insanely inefficient and gives a rather poor user experience, as every single interaction feels laggy, making the whole app feel a bit cheap.

To remedy this, we will now use the `Repository`, a new way of handling changes in collections and synchronization with the Cloud. This makes every interaction instantaneous and gives a much much better user experience, while still assuring a 100% synchronization with the Cloud.

We will use @TuanDangNguyen2003's Memento as an example for the following.

## How to use it
### `Repository` setup
The `Repository` is simply a controlled collection. We add, update and remove documents from the `Repository` using a `handler`. A `Repository` also has a unique key, different from the collection path, that we use to store it locally.

In Memento, the primary collection is `decks` that are of type `Memento.Deck`, so we write the following :
```typescript
const DecksRepository = createRepository<Memento.Deck>("my-decks");
```
You can see here that `my-decks` is the unique key of this repository.

We are working in the `app/(app)/deck` directory and I want the all the pages that are inside of it to be able to access the `Repository` and manipulate it.
I will then create a new file `app/(app)/deck/_layout.tsx` that will simply propagate the whole repository to all the files inside the directory. I will also put the repository creation code there :
```typescript
export const DecksRepository = createRepository<Memento.Deck>("my-decks");

const MementoLayout: ApplicationLayout = () => {
    const { uid } = useAuth();
    const collection = CollectionOf<Memento.Deck>(`users/${uid}/decks`);
    return <RepositoryLayout repository={DecksRepository} collection={collection} />;
};

export default MementoLayout;

```
As you see in this example, I link the local repository `my-decks` to the user's deck collection in Firebase.

### Accessing the `Repository` inside the routes
There are two main hooks to access the repository from inside the routes :

The `useRepository` hook gives you access to all the documents in the `Repository` as well as the `handler`.
```typescript
const [decks, handler] = useRepository(DecksRepository);
```

The `useRepositoryDocument` hook directly gives you a specific document in the Repository, as well as the `handler`.
```typescript
const [deck, handler] = useRepositoryDocument(id, DecksRepository);
```

### Manipulating the `Repository` using the `handler`.
The `handler` is an object that gives you full control of the repository, while still assuring that everything is kept in sync with the Cloud.

#### Adding a document
To add new documents, we will use `handler.addDocument(data, idSupplier)`.
`idSupplier` here is expected to be a Firebase Cloud Function call (with `callFunction` of course) that, as a return value, has the `id` of the newly formed document.
```typescript
const deck = {
    name: deckName,
    cards: []
}

handler.addDocument(deck, callFunction(Memento.Functions.createDeck, { deck }));
```
We do **NOT** use the `await` keyword as we want it to stay a `Promise`.

These lines effectively allow us to *optimistically* update the UI without waiting for the function call to end. Of course if the function call fails, it reverts back and removes the document, but it should be a rare occurrence. If you use the `handler`, you're the one in charge to assure that **failing** is a very rare occurrence (by verifying the inputs before calling the function for example).

#### Modifying a document
To modify an existing document (like updating one of its fields), we will use `handler.modifyDocument(id, partialData, syncCallback)`.
For `partialData`, you only put the fields that you want to modify, everything else remains intact.
```typescript
const newDeckName = "My amazing deck";

handler.modifyDocument(deckId, {name: newDeckName}, (id) => {
    callFunction(Memento.Functions.updateDeck, { deckId: id, deck: {name: newDeckName} });
});
```
As you see in this example, `syncCallback` supplies you with the real `id` of the document. For technical reasons, you can't directly use `deckId` (or at least you won't have the guaranty that you modify what you think you're modifying, the full implementation is in `@/hooks/repository`).

#### Deleting a document
To delete a document, we will use `handler.deleteDocument(id, syncCallback)` :
```typescript
handler.deleteDocument(id, (id) => {
    callFunction(Memento.Functions.deleteDecks, { deckIds: [id] });
});
```
Or delete multiple documents with `handler.deleteDocuments(ids, syncCallback)` :
```typescript
handler.deleteDocuments(deckIds, (ids) => {
    callFunction(Memento.Functions.deleteDecks, { deckIds: ids });
});
```

## Other minor improvements
#### Type safe route parameters (`utils/routeParameters.ts`)
Are you tired of checking that `id` is a `string` a not a `string[]` when you use `useLocalSearchParams()` ? You can now use `useStringParameters()` instead ! (it does the same thing, but just has the right signature)

You can now also instantiate a generic `ApplicationRouteParameters<T>` to use with `pushWithParameters(routeParameters, params)` and `useRouteParameters(routeParameters)`. This gives you full type safety when you pass multiple parameters so that you can feel confident about your code ! This also removes trivial errors like renaming in one place but forgetting in another.
#### Common mocking namespaces
One of the ways testing is painful is to find the mocks for the components and functions we use, please checkout the most common mocks at `__test__/__mocks__`. They aren't that many for now but please feel free to add all the ones that you find yourself using often!

FerdinandGomerChiu reviewed (commented):
Hi, this is a lot of work! Before I even dive into the review, I want to pint out a few things.

SonarCloud detects some issues with your code, namely about reliability (C grading) and a detail about the use of a random number generator in edweiss-app/utils/uid.ts : 
``` typescript
export default function generateUID(length: number) {

	const alphabet = "qwertyuiopasdfghjklzxcvbnQWERTYUIOPASDFGHJKLZXCVBN0123456789";

	let res = "";

	for (let i = 0; i < length; i++) {

		res += alphabet.charAt(Math.floor(Math.random()
 * alphabet.length));


	}

	return res;

}
```
How do you plan on solving these issues?

I will add other comments step by step as I'm reading your changes.

FerdinandGomerChiu commented on edweiss-app/hooks/repository.tsx:
```diff
@@ -0,0 +1,348 @@
+	 * @param data The data of the document to be added optimistically.
+	 * @param idSupplier Typically a `callFunction` with an `{ id: string }` as a success return value.
+	 */
+	readonly addDocument: (data: T, idSupplier: Promise<CallResult<{ id: string }, unknown>>) => void,
```
I see that there are two different functions to add or update a document. I can see an issue with this: the DocumentReference<T>.set() function that we used in createQuizAttempt acted as both a creation tool (when there are no attempts at the designated path) or a modifying tool (when there was indeed an attempt). 
```typescript
export const createQuizAttempt = onSanitizedCall(QuizzesAttempts.Functions.createQuizAttempt, {
	courseId: Predicate.isNonEmptyString,
	path: Predicate.isNonEmptyString,
	quizId: Predicate.isNonEmptyString,
	quizAttempt: CustomPredicateQuiz.isValidQuizAttempt
}, async (userId, args) => {
	const ref = getDocumentRef(CollectionOf<QuizzesAttempts.QuizAttempt>(args.path), userId);

	await ref.set(args.quizAttempt);
	console.log("Document successfully updated!");
	return ok({ id: ref.id });
});
```
Do we now have to decouple those two cases?

holoisme replied:
No, as the Quiz is not applicable in the case of the Repository.
Repository is only used for modules that are actively under the CRUD model, for example it can be used for :
 - Flashcards
 - Todos
 - Notifications
Quizzes and attempts do not fit the Repository purpose.

FerdinandGomerChiu replied:
Ahh, I see, thank you. Should the other members be immediately aware of the changes you made and use the handler this sprint, or do you believe the code is easily refactorable later?

holoisme replied:
I encourage them to adopt it as early as possible, because 1. it will inevitably provide the best possible user experience, and 2. the tests do change a bit so it's better to make the transition quickly.

FerdinandGomerChiu replied:
Since your PR is not merged yet, I propose we refactor right after tomorrows meeting. Do you agree?

holoisme replied:
It's as you guys want, I'm of course not forcing anyone to use it. Only people who want their module to be blazingly fast, responsive and pleasant should implement it.

FerdinandGomerChiu reviewed (commented):
I see that you changed a few test files by replacing space characters with tabs: I believe this is unintended and is a consequence of your IDE settings. Please change them so that spaces are not automatically changed; you can still use the setting for your own code of course.
![image](https://github.com/user-attachments/assets/4e4bab1c-d1ea-4e65-8d0e-47e80862c041)

holoisme commented:
> I see that you changed a few test files by replacing space characters with tabs: I believe this is unintended and is a consequence of your IDE settings. Please change them so that spaces are not automatically changed; you can still use the setting for your own code of course. ![image](https://private-user-images.githubusercontent.com/162486036/388509256-4e4bab1c-d1ea-4e65-8d0e-47e80862c041.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIxOTI0MTAsIm5iZiI6MTczMjE5MjExMCwicGF0aCI6Ii8xNjI0ODYwMzYvMzg4NTA5MjU2LTRlNGJhYjFjLWQxZWEtNGU2NS04ZDBlLTQ3ZTgwODYyYzA0MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyMVQxMjI4MzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kNTU2Y2VmZGJhZWI1OTVmZDkzYWNkMzRjMzUyMGJkYjI4MWUxYmZjZGI0MTYxZjM2NzI2YWI2ODlhZWI0ZmMzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.7UdZ5xbVr0W0R_b1eNh-BdBEYRwwUbyfTdYUWnCgu8Y)

I did not only change the tabs into spaces, I added the repository mock to help @TuanDangNguyen2003 refactor.
Tabs are industry standard when working with HTML-like syntax like JSX in our case ;)

holoisme commented:
> Hi, this is a lot of work! Before I even dive into the review, I want to pint out a few things.
> 
> SonarCloud detects some issues with your code, namely about reliability (C grading) and a detail about the use of a random number generator in edweiss-app/utils/uid.ts :
> 
> ```ts
> export default function generateUID(length: number) {
> 
> 	const alphabet = "qwertyuiopasdfghjklzxcvbnQWERTYUIOPASDFGHJKLZXCVBN0123456789";
> 
> 	let res = "";
> 
> 	for (let i = 0; i < length; i++) {
> 
> 		res += alphabet.charAt(Math.floor(Math.random()
>  * alphabet.length));
> 
> 
> 	}
> 
> 	return res;
> 
> }
> ```
> 
> How do you plan on solving these issues?
> 
> I will add other comments step by step as I'm reading your changes.

Can you add the link to this specific issue ? I have trouble finding it.
And in any case, we now know that sonarcloud can be bad at recognizing intentions of usage (with the infamous hook after return), so I don't pay much attention to it.

FerdinandGomerChiu commented:
> > I see that you changed a few test files by replacing space characters with tabs: I believe this is unintended and is a consequence of your IDE settings. Please change them so that spaces are not automatically changed; you can still use the setting for your own code of course. ![image](https://private-user-images.githubusercontent.com/162486036/388509256-4e4bab1c-d1ea-4e65-8d0e-47e80862c041.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIxOTI0MTAsIm5iZiI6MTczMjE5MjExMCwicGF0aCI6Ii8xNjI0ODYwMzYvMzg4NTA5MjU2LTRlNGJhYjFjLWQxZWEtNGU2NS04ZDBlLTQ3ZTgwODYyYzA0MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyMVQxMjI4MzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kNTU2Y2VmZGJhZWI1OTVmZDkzYWNkMzRjMzUyMGJkYjI4MWUxYmZjZGI0MTYxZjM2NzI2YWI2ODlhZWI0ZmMzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.7UdZ5xbVr0W0R_b1eNh-BdBEYRwwUbyfTdYUWnCgu8Y)
> 
> I did not only change the tabs into spaces, I added the repository mock to help @TuanDangNguyen2003 refactor. Tabs are industry standard when working with HTML-like syntax like JSX in our case ;)

Yes, of course! This was a great addition. You are also right that this is the industry standard: just keep in mind that not all of us keep the standard unfortunately :( and your changes might introduce conflicts. Have you checked with @TuanDangNguyen2003 if he changed those files during this week's sprint?

holoisme commented:
> > > I see that you changed a few test files by replacing space characters with tabs: I believe this is unintended and is a consequence of your IDE settings. Please change them so that spaces are not automatically changed; you can still use the setting for your own code of course. ![image](https://private-user-images.githubusercontent.com/162486036/388509256-4e4bab1c-d1ea-4e65-8d0e-47e80862c041.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIxOTI0MTAsIm5iZiI6MTczMjE5MjExMCwicGF0aCI6Ii8xNjI0ODYwMzYvMzg4NTA5MjU2LTRlNGJhYjFjLWQxZWEtNGU2NS04ZDBlLTQ3ZTgwODYyYzA0MS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyMVQxMjI4MzBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kNTU2Y2VmZGJhZWI1OTVmZDkzYWNkMzRjMzUyMGJkYjI4MWUxYmZjZGI0MTYxZjM2NzI2YWI2ODlhZWI0ZmMzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.7UdZ5xbVr0W0R_b1eNh-BdBEYRwwUbyfTdYUWnCgu8Y)
> > 
> > 
> > I did not only change the tabs into spaces, I added the repository mock to help @TuanDangNguyen2003 refactor. Tabs are industry standard when working with HTML-like syntax like JSX in our case ;)
> 
> Yes, of course! This was a great addition. You are also right that this is the industry standard: just keep in mind that not all of us keep the standard unfortunately :( and your changes might introduce conflicts. Have you checked with @TuanDangNguyen2003 if he changed those files during this week's sprint?

I was just kidding of course, conflicts can be a pain, you are a 100% right. I just changed the setting.

FerdinandGomerChiu commented:
> > Hi, this is a lot of work! Before I even dive into the review, I want to pint out a few things.
> > SonarCloud detects some issues with your code, namely about reliability (C grading) and a detail about the use of a random number generator in edweiss-app/utils/uid.ts :
> > ```ts
> > export default function generateUID(length: number) {
> > 
> > 	const alphabet = "qwertyuiopasdfghjklzxcvbnQWERTYUIOPASDFGHJKLZXCVBN0123456789";
> > 
> > 	let res = "";
> > 
> > 	for (let i = 0; i < length; i++) {
> > 
> > 		res += alphabet.charAt(Math.floor(Math.random()
> >  * alphabet.length));
> > 
> > 
> > 	}
> > 
> > 	return res;
> > 
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > How do you plan on solving these issues?
> > I will add other comments step by step as I'm reading your changes.
> 
> Can you add the link to this specific issue ? I have trouble finding it. And in any case, we now know that sonarcloud can be bad at recognizing intentions of usage (with the infamous hook after return), so I don't pay much attention to it.

This is the link: you can access the report by clicking on "SonarCloud analysis details", at the bottom of your PR. Then click on "Security hotspot".

https://sonarcloud.io/project/security_hotspots?id=SnowWhiteEPFL_edweiss&branch=feature%2Frepository&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true

I am aware that SonarCloud is nitpicky or downright wrong at times. I am just worried about grading. Did we already talk about this with the coaches?

holoisme commented:
> > > Hi, this is a lot of work! Before I even dive into the review, I want to pint out a few things.
> > > SonarCloud detects some issues with your code, namely about reliability (C grading) and a detail about the use of a random number generator in edweiss-app/utils/uid.ts :
> > > ```ts
> > > export default function generateUID(length: number) {
> > > 
> > > 	const alphabet = "qwertyuiopasdfghjklzxcvbnQWERTYUIOPASDFGHJKLZXCVBN0123456789";
> > > 
> > > 	let res = "";
> > > 
> > > 	for (let i = 0; i < length; i++) {
> > > 
> > > 		res += alphabet.charAt(Math.floor(Math.random()
> > >  * alphabet.length));
> > > 
> > > 
> > > 	}
> > > 
> > > 	return res;
> > > 
> > > }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > How do you plan on solving these issues?
> > > I will add other comments step by step as I'm reading your changes.
> > 
> > 
> > Can you add the link to this specific issue ? I have trouble finding it. And in any case, we now know that sonarcloud can be bad at recognizing intentions of usage (with the infamous hook after return), so I don't pay much attention to it.
> 
> This is the link: you can access the report by clicking on "SonarCloud analysis details", at the bottom of your PR. Then click on "Security hotspot".
> 
> https://sonarcloud.io/project/security_hotspots?id=SnowWhiteEPFL_edweiss&branch=feature%2Frepository&issueStatuses=OPEN,CONFIRMED&sinceLeakPeriod=true
> 
> I am aware that SonarCloud is nitpicky or downright wrong at times. I am just worried about grading. Did we already talk about this with the coaches?

Yup I just saw this issue, using a pseudorandom generator is 100% safe here as it only provides temporary fake uids for repository documents when they're added.
Also yes I believe we did, we only care about coverage.

holoisme commented on edweiss-app/__test__/__mocks__/index.tsx:
```diff
@@ -0,0 +1,29 @@
+ * To refactor because those aren't working, we'll see.
+ */
+
+export namespace CommonMock {
```
Lol, very annoying that we can't have openly temporary things for everyone to fix.

FerdinandGomerChiu commented on edweiss-app/hooks/routeParameters.ts:
```diff
@@ -0,0 +1,48 @@
+ * 
+ * @returns The parsed parameters.
+ */
+export function useRouteParameters<T>(_: ApplicationRouteSignature<T>): T {
```
Does this allow us to push whatever object we want? Is there a limit to stringifying?

holoisme replied:
Well you can push anything that you would have been able to with the standard `router.push`. Meaning every JSON compatible value (strings, numbers, lists, nested objects) are allowed, but functions aren't.

TuanDangNguyen2003 commented on edweiss-app/__test__/__mocks__/repository.tsx:
I am currently recheck the mocking of repository because we encounter an "undefined" component because a mocking doesn't work well. I am taking a deep look at this

TuanDangNguyen2003 reviewed (commented):
It is a very big change, thank you for all the hard works. Undoubtedly this will enhance the smoothness and quickness of the app and this will provide a great experience for user. Overall your files look good, are well organized and clear. However just a small note about mocking of repository because for now it doesn't work. I am also trying to make the tests run.

FerdinandGomerChiu reviewed (approved):
This PR is well documented. The conversations cleared the doubts I had, and I believe this change will allow to work much faster. In fact, I need the changes the earliest possible, to better work on quiz creation. 
Thank you! Keep it up!

TuanDangNguyen2003 commented:
Ok I have just finished fixing the tests so that they run without problem. In general, __mocks__/repository.tsx is good but in order to use for example modifyDocument, ... we must mock them specifically with real hard coded data in order for the test to have an actual value.

Other than that, you can merge in the main :D
